pkg/version/version_parser.go

```diff
@@ -99,34 +145,66 @@ func (psv ParsedSemVer) Less(other ParsedSemVer) bool {
 		return psv.patch < other.patch
 	}
 
-	// last resort check if one is prereleas and the other isn't
+	// compare prerelease strings as major.minor.patch are equal
+	return psv.comparePrerelease(other)
+}
+
+// comparePrerelease compares the prerelease part of 2 ParsedSemVer objects
+// the return value must conform to psv.prerelease < other.prerelease following comparison rules from https://semver.org/
+func (psv ParsedSemVer) comparePrerelease(other ParsedSemVer) bool {
+	// last resort before parsing prerelease: check if one is prerelease and the other isn't
 	if psv.prerelease != "" && other.prerelease == "" {
 		return true
 	}
 
-	return false
-}
+	if psv.prerelease == "" && other.prerelease != "" {
+		return false
+	}
 
-func (psv ParsedSemVer) GetPreviousMinor() (*ParsedSemVer, error) {
-	major := psv.Major()
-	minor := psv.Minor()
+	// tokenize prereleases and compare them
+	prereleaseTokens := strings.Split(psv.prerelease, prereleaseTokenSeparator)
+	otherPrereleaseTokens := strings.Split(other.prerelease, prereleaseTokenSeparator)
 
-	if minor > 0 {
-		// We have at least one previous minor version in the current
-		// major version series. Set the patch to zero to guarnatee the
-		// version exists, the number of patch releases varies.
-		return NewParsedSemVer(major, minor-1, 0, psv.Prerelease(), psv.BuildMetadata()), nil
+	// compute the min amount of tokens
+	minPrereleaseTokens := len(prereleaseTokens)
+	if len(otherPrereleaseTokens) < minPrereleaseTokens {
+		minPrereleaseTokens = len(otherPrereleaseTokens)
 	}
 
-	// We are at the first minor of the current major version series. To
-	// figure out the previous minor, we need to rely on knowledge of
-	// the release versions from the past major series'.
-	switch major {
-	case 8:
-		return NewParsedSemVer(7, 17, 10, psv.Prerelease(), psv.BuildMetadata()), nil
+	for i := 0; i < minPrereleaseTokens; i++ {
```

Comment:
What problem is this solving? Do we need this code to handle something in our current or planned release versioning scheme or is it just in case?

Reply:
This is the correct implementation for comparing 2 semver version strings. Prerelease strings should be compared token by token as strings or numbers depending on their format.

Right now we only use "SNAPSHOT" prerelease string and the previous impl would just compare 2 versions based on the presence of prerelease string without comparing those, leading to versions like
`1.2.3-SNAPSHOT` , `1.2.3-alpha`, `1.2.3-SNAPHOT.extraprereleasetoken` not to have an definite order.

This new implementation is able to parse the tokens and correctly determine precedence based on the prerelease token: an example when this is gonna come into play is of we decide to use repackaged agents in upgrade tests as `1.2.3-SNAPSHOT.repackaged` now has a defined order when compared with `1.2.3-SNAPSHOT`

This is a part of a general overhaul of agent version handling correct versioning both at packaging and runtime which I then extracted to put in a separate PR as it would make this change too big.